components/raftstore/src/store/fsm/store.rs

```diff
@@ -683,61 +647,36 @@ impl<EK: KvEngine, ER: RaftEngine, T: Transport> RaftPoller<EK, ER, T> {
         // Only enable the fail point when the store id is equal to 3, which is
         // the id of slow store in tests.
         fail_point!("on_raft_ready", self.poll_ctx.store_id() == 3, |_| {});
-        if self.poll_ctx.trans.need_flush()
-            && (!self.poll_ctx.kv_wb.is_empty() || !self.poll_ctx.raft_wb.is_empty())
-        {
-            self.poll_ctx.trans.flush();
-        }
-        let ready_cnt = self.poll_ctx.ready_res.len();
-        self.poll_ctx.raft_metrics.ready.has_ready_region += ready_cnt as u64;
-        fail_point!("raft_before_save");
-        if !self.poll_ctx.kv_wb.is_empty() {
-            let mut write_opts = WriteOptions::new();
-            write_opts.set_sync(true);
-            self.poll_ctx
-                .kv_wb
-                .write_opt(&write_opts)
-                .unwrap_or_else(|e| {
-                    panic!("{} failed to save append state result: {:?}", self.tag, e);
-                });
-            let data_size = self.poll_ctx.kv_wb.data_size();
-            if data_size > KV_WB_SHRINK_SIZE {
-                self.poll_ctx.kv_wb = self.poll_ctx.engines.kv.write_batch_with_cap(4 * 1024);
-            } else {
-                self.poll_ctx.kv_wb.clear();
+        if let Some(mut write_worker) = self.poll_ctx.write_worker.take() {
+            if !write_worker.batch.is_empty() {
+                self.poll_ctx.trans.flush();
             }
-        }
-        fail_point!("raft_between_save");
-
-        if !self.poll_ctx.raft_wb.is_empty() {
-            fail_point!(
-                "raft_before_save_on_store_1",
-                self.poll_ctx.store_id() == 1,
-                |_| {}
-            );
-            self.poll_ctx
-                .engines
-                .raft
-                .consume_and_shrink(
-                    &mut self.poll_ctx.raft_wb,
-                    true,
-                    RAFT_WB_SHRINK_SIZE,
-                    4 * 1024,
-                )
-                .unwrap_or_else(|e| {
-                    panic!("{} failed to save raft append result: {:?}", self.tag, e);
-                });
-        }
-
-        self.poll_ctx.perf_context.report_metrics();
-        fail_point!("raft_after_save");
-        if ready_cnt != 0 {
-            let mut ready_res = mem::take(&mut self.poll_ctx.ready_res);
-            for ready in ready_res.drain(..) {
-                PeerFsmDelegate::new(&mut peers[ready.batch_offset], &mut self.poll_ctx)
-                    .post_raft_ready_append(ready);
+            let mut readies = write_worker.write_to_db(false);
+            self.poll_ctx.write_worker = Some(write_worker);
+
+            for fsm in peers {
```

Comment:
It's not as efficient as past. In the past, only readiness are looped. For example, if there a lot of heartbeats, there can be a lot of waste.

Reply:
I added `pos` for `handle_normal` to fix this issue. I think it's clearer than `processed_fsm_count`.